File: q0864/q0864.cpp
#include <vector>
#include <string>
#include <set>
#include <queue>
using namespace std;

class Solution
{
public:
    int shortestPathAllKeys(vector<string> &grid)
    {
        int m = grid.size();
        int n = grid[0].size();
        set<int> keySet;
        set<int> lockSet;
        int startRow = -1;
        int startCol = -1;
        int keyState = 0;

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                char c = grid[i][j];
                // a key if found => add to ket set
                if (c >= 'a' && c <= 'f')
                {
                    keyState += (1 << (c - 'a'));
                    keySet.insert(c);
                    // a lock is found => add to lock set
                }
                else if (c >= 'A' && c <= 'F')
                {
                    lockSet.insert(c);
                    // determine the beginning
                }
                else if (c == '@')
                {
                    startRow = i;
                    startCol = j;
                }
            }
        }

        vector<pair<int, int>> moves = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        int initialKeyState = 0;
        set<vector<int>> visited;
        visited.insert({initialKeyState, startRow, startCol});

        queue<vector<int>> q;
        q.push({startRow, startCol, initialKeyState, 0});
        while (!q.empty())
        {
            vector<int> currState = q.front();
            q.pop();
            int currRow = currState[0];
            int currCol = currState[1];
            int currKeyState = currState[2];
            int currPath = currState[3];
            for (auto move : moves)
            {
                int newRow = currRow + move.first;
                int newCol = currCol + move.second;
                if (newRow >= 0 && newRow < m && newCol >= 0 && newCol < n && grid[newRow][newCol] != '#')
                {
                    char c = grid[newRow][newCol];
                    if (keySet.find(c) != keySet.end())
                    {
                        // If the key has been collected before, there is not need to explore
                        if (((1 << (c - 'a')) & currKeyState) != 0)
                            continue;

                        // Collect the key and check if we have all the keys
                        int newKeyState = (currKeyState | (1 << (c - 'a')));

                        // If we have all the keys, return the path to obtain the last key
                        if (newKeyState == keyState)
                            return currPath + 1;

                        // If we do not have all the keys, generate a new state, add to visited
                        visited.insert({newKeyState, newRow, newCol});
                        q.push({newRow, newCol, newKeyState, currPath + 1});
                    }

                    // If we have a lock, but do not have a corresponding key, we cannot explore it
                    if ((lockSet.find(c) != lockSet.end()) && ((currKeyState & (1 << (c - 'A'))) == 0))
                    {
                        continue;
                    }

                    // If we do not see lock / key, it is a regular cell, there will not be a new keyState
                    // Check if we have already had this combination on the queue, if not, add it.
                    if (visited.find({currKeyState, newRow, newCol}) == visited.end())
                    {
                        visited.insert({currKeyState, newRow, newCol});
                        q.push({newRow, newCol, currKeyState, currPath + 1});
                    }
                }
            }
        }

        return -1;
    }
};

File: q0864/q0864.java
package q0864;

import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

class Solution {
    public int shortestPathAllKeys(String[] grid) {
        int m = grid.length;
        int n = grid[0].length();

        Set<Character> keySet = new HashSet<>();
        Set<Character> lockSet = new HashSet<>();
        int startCol = -1;
        int startRow = -1;
        int keyState = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                char c = grid[i].charAt(j);
                // determine if c is a key
                if (c >= 'a' && c <= 'f') {
                    keyState += (1 << (c - 'a'));
                    keySet.add(c);
                }
                // determine if c is a lock
                if (c >= 'A' && c <= 'F') {
                    lockSet.add(c);
                }
                // determine starting position
                if (c == '@') {
                    startRow = i;
                    startCol = j;
                }
            }
        }

        Map<Integer, Set<Pair<Integer, Integer>>> visited = new HashMap<>();
        int initialKeyState = 0;
        visited.put(initialKeyState, new HashSet<>());
        visited.get(initialKeyState).add(new Pair<>(startRow, startCol));

        int[][] moves = new int[][] { { 0, 1 }, { 1, 0 }, { -1, 0 }, { 0, -1 } };

        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[] { startRow, startCol, initialKeyState, 0 });
        while (!queue.isEmpty()) {
            int[] currentState = queue.poll();
            int currRow = currentState[0];
            int currCol = currentState[1];
            int currKeyState = currentState[2];
            int currPath = currentState[3];
            for (int[] move : moves) {
                int newRow = currRow + move[0];
                int newCol = currCol + move[1];
                if (newRow >= 0 && newRow < m && newCol >= 0 && newCol < n && grid[newRow].charAt(newCol) != '#') {
                    char c = grid[newRow].charAt(newCol);

                    if (keySet.contains(c)) {
                        if (((1 << (c - 'a')) & currKeyState) != 0)
                            continue;
                        int newKeyState = (currKeyState | (1 << (c - 'a')));
                        if (newKeyState == keyState)
                            return currPath + 1;
                        visited.putIfAbsent(newKeyState, new HashSet<>());
                        visited.get(newKeyState).add(new Pair<>(newRow, newCol));
                        queue.add(new int[] { newRow, newCol, newKeyState, currPath + 1 });
                    }

                    if (lockSet.contains(c) && ((currKeyState & (1 << (c - 'A'))) == 0)) {
                        continue;
                    }

                    if (!visited.get(currKeyState).contains(new Pair<>(newRow, newCol))) {
                        visited.get(currKeyState).add(new Pair<>(newRow, newCol));
                        queue.add(new int[] { newRow, newCol, currKeyState, currPath + 1 });
                    }
                }
            }
        }
        return -1;
    }
}
